File: web/services/owner/payment.service.ts
import api from '@/services/api';
import type { Payment, PaymentFilters, ConfirmCashPaymentRequest, PaymentReceipt } from '@/types/payment.types';
import type { ApiResponse, PaginatedResponse } from '@/types/api.types';

export const getPayments = async (filters?: PaymentFilters, page = 1, limit = 20): Promise<PaginatedResponse<Payment>> => {
  const { data } = await api.get<PaginatedResponse<Payment>>('/owner/payments', {
    params: { ...filters, page, limit },
  });
  return data;
};

export const confirmCashPayment = async (paymentData: ConfirmCashPaymentRequest): Promise<Payment> => {
  const { data } = await api.post<ApiResponse<Payment>>('/owner/payments/confirm-cash', paymentData);
  return data.data;
};

export const getReceipt = async (paymentId: number): Promise<PaymentReceipt> => {
  const { data } = await api.get<ApiResponse<PaymentReceipt>>(`/owner/payments/${paymentId}/receipt`);
  return data.data;
};

export const downloadReceipt = async (paymentId: number): Promise<Blob> => {
  const { data } = await api.get<Blob>(`/owner/payments/${paymentId}/receipt/pdf`, {
    responseType: 'blob',
  });
  return data;
};

File: mobile/lib/services/payment_service.dart
import '../models/payment.dart';
import 'api_client.dart';

class PaymentService {
  final _apiClient = ApiClient();

  Future<List<Payment>> getPayments({
    String? status,
    int? page,
    int? limit,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (status != null) queryParams['status'] = status;
      if (page != null) queryParams['page'] = page;
      if (limit != null) queryParams['limit'] = limit;

      final response = await _apiClient.get(
        '/payments',
        queryParameters: queryParams,
      );

      final data = response.data as Map<String, dynamic>;
      final paymentsList = data['payments'] as List;
      return paymentsList
          .map((item) => Payment.fromJson(item as Map<String, dynamic>))
          .toList();
    } catch (e) {
      rethrow;
    }
  }

  Future<Payment> getPaymentById(String paymentId) async {
    try {
      final response = await _apiClient.get('/payments/$paymentId');
      final data = response.data as Map<String, dynamic>;
      return Payment.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<Payment> createPayment({
    required String billId,
    required double amount,
    required String method,
    required DateTime paymentDate,
    String? proofImage,
    String? notes,
  }) async {
    try {
      final response = await _apiClient.post(
        '/payments',
        data: {
          'billId': billId,
          'amount': amount,
          'method': method,
          'paymentDate': paymentDate.toIso8601String(),
          'proofImage': proofImage,
          'notes': notes,
        },
      );

      final data = response.data as Map<String, dynamic>;
      return Payment.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<Payment> confirmPayment({
    required String paymentId,
    required String status,
    String? notes,
  }) async {
    try {
      final response = await _apiClient.post(
        '/payments/$paymentId/confirm',
        data: {
          'status': status,
          'notes': notes,
        },
      );

      final data = response.data as Map<String, dynamic>;
      return Payment.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<String> getReceiptUrl(String paymentId) async {
    try {
      final response = await _apiClient.get('/payments/$paymentId/receipt');
      final data = response.data as Map<String, dynamic>;
      return data['url'] as String;
    } catch (e) {
      rethrow;
    }
  }

  Future<Map<String, dynamic>> getPaymentStats() async {
    try {
      final response = await _apiClient.get('/payments/stats');
      return response.data as Map<String, dynamic>;
    } catch (e) {
      rethrow;
    }
  }

  // Owner: Confirm cash payment
  Future<Payment> confirmCashPayment({
    required int billId,
  }) async {
    try {
      final response = await _apiClient.post(
        '/owner/payments/confirm-cash',
        data: {'billId': billId},
      );
      final data = response.data as Map<String, dynamic>;
      return Payment.fromJson(data['data'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to confirm cash payment: $e');
    }
  }
}
